File: backend/app/services/severity.py
from __future__ import annotations

import re
from typing import Iterable


CANONICAL_SEVERITY_LABELS = ("critical", "high", "medium", "low", "info")

KNOWN_EXPLOITED_MARKERS = {
    "knownexploited",
    "knownexploitedvulnerability",
    "knownexploitedvulnerabilities",
    "kev",
    "cisakev",
    "已知利用",
    "已知被利用",
    "已知存在利用",
    "已知有利用",
}

KNOWN_EXPLOITED_QUERY_VALUES = {
    "known_exploited",
    "known-exploited",
    "known exploited",
    "known exploited vulnerability",
    "known exploited vulnerabilities",
    "kev",
    "cisa kev",
    "cisa-kev",
    "已知利用",
    "已知被利用",
    "已知存在利用",
    "已知有利用",
}

EMPTY_SEVERITY_MARKERS = {
    "unknown",
    "unk",
    "none",
    "null",
    "na",
    "n/a",
    "-",
    "未知",
    "不详",
    "暂无",
    "无",
}

SEVERITY_ALIASES: dict[str, tuple[str, ...]] = {
    "critical": (
        "critical",
        "crit",
        "severe",
        "serious",
        "严重",
        "超危",
        "严重 / critical",
        "critical / 严重",
        "严重critical",
        "critical严重",
    ),
    "high": (
        "high",
        "important",
        "高",
        "高危",
        "高风险",
        "highrisk",
        "高危 / high",
        "high / 高危",
        "高危high",
        "high高危",
    ),
    "medium": (
        "medium",
        "moderate",
        "中",
        "中危",
        "中风险",
        "mediumrisk",
        "中危 / medium",
        "medium / 中危",
        "中危medium",
        "medium中危",
    ),
    "low": (
        "low",
        "低",
        "低危",
        "低风险",
        "lowrisk",
        "低危 / low",
        "low / 低危",
        "低危low",
        "low低危",
    ),
    "info": (
        "info",
        "informational",
        "information",
        "信息",
        "提示",
        "信息 / info",
        "info / 信息",
        "信息info",
        "info信息",
    ),
}

_ALIAS_TO_CANONICAL = {
    alias: canonical
    for canonical, aliases in SEVERITY_ALIASES.items()
    for alias in aliases
}


def normalize_severity_label(value: object | None) -> str | None:
    text = _clean_text(value)
    if text is None:
        return None

    key = _compact_key(text)
    if (
        key in KNOWN_EXPLOITED_MARKERS
        or key in EMPTY_SEVERITY_MARKERS
        or text.lower() in EMPTY_SEVERITY_MARKERS
    ):
        return None
    return _ALIAS_TO_CANONICAL.get(key, text.lower())


def is_known_exploited_marker(value: object | None) -> bool:
    text = _clean_text(value)
    return bool(text and _compact_key(text) in KNOWN_EXPLOITED_MARKERS)


def severity_query_values(labels: Iterable[object]) -> set[str]:
    values: set[str] = set()
    for label in labels:
        text = _clean_text(label)
        if text is None or is_known_exploited_marker(text):
            continue

        normalized = normalize_severity_label(text)
        if normalized is None:
            continue

        aliases = SEVERITY_ALIASES.get(normalized)
        if aliases is None:
            values.add(text.lower())
            continue

        values.update(alias.lower() for alias in aliases)
        values.add(normalized)
    return values


def _clean_text(value: object | None) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    return text or None


def _compact_key(value: str) -> str:
    return re.sub(r"[\s/_\-.]+", "", value.strip().lower())

File: frontend/src/utils/severity.ts
import { t } from "@/app/i18n";
export const severityOptions = [
  { label: t("严重 / Critical"), value: "critical" },
  { label: t("高危 / High"), value: "high" },
  { label: t("中危 / Medium"), value: "medium" },
  { label: t("低危 / Low"), value: "low" },
  { label: t("信息 / Info"), value: "info" }
];

const severityLabels: Record<string, string> = {
  critical: t("严重 / Critical"),
  high: t("高危 / High"),
  medium: t("中危 / Medium"),
  low: t("低危 / Low"),
  info: t("信息 / Info")
};

const severityColors: Record<string, string> = {
  critical: "red",
  high: "volcano",
  medium: "orange",
  low: "blue",
  info: "default"
};

const severityAliases: Record<string, string> = {
  critical: "critical",
  crit: "critical",
  severe: "critical",
  serious: "critical",
  严重: "critical",
  超危: "critical",
  严重critical: "critical",
  critical严重: "critical",
  high: "high",
  important: "high",
  高: "high",
  高危: "high",
  高风险: "high",
  highrisk: "high",
  高危high: "high",
  high高危: "high",
  medium: "medium",
  moderate: "medium",
  中: "medium",
  中危: "medium",
  中风险: "medium",
  mediumrisk: "medium",
  中危medium: "medium",
  medium中危: "medium",
  low: "low",
  低: "low",
  低危: "low",
  低风险: "low",
  lowrisk: "low",
  低危low: "low",
  low低危: "low",
  info: "info",
  informational: "info",
  information: "info",
  信息: "info",
  提示: "info",
  信息info: "info",
  info信息: "info"
};

const knownExploitedMarkers = new Set([
  "knownexploited",
  "knownexploitedvulnerability",
  "knownexploitedvulnerabilities",
  "kev",
  "cisakev",
  t("已知利用"),
  t("已知被利用"),
  t("已知存在利用"),
  t("已知有利用")
]);

const emptySeverityMarkers = new Set([
  "unknown",
  "unk",
  "none",
  "null",
  "na",
  "n/a",
  "-",
  t("未知"),
  t("不详"),
  t("暂无"),
  t("无")
]);

export function normalizeSeverityLabel(value?: string | null) {
  const trimmed = value?.trim();
  if (!trimmed) {
    return null;
  }
  const key = compactKey(trimmed);
  if (
    knownExploitedMarkers.has(key) ||
    emptySeverityMarkers.has(key) ||
    emptySeverityMarkers.has(trimmed.toLowerCase())
  ) {
    return null;
  }
  return severityAliases[key] ?? trimmed.toLowerCase();
}

export function isKnownExploitedSeverity(value?: string | null) {
  const trimmed = value?.trim();
  return Boolean(trimmed && knownExploitedMarkers.has(compactKey(trimmed)));
}

export function severityDisplayValue(value?: string | null) {
  const normalized = normalizeSeverityLabel(value);
  return normalized ? severityLabels[normalized] ?? normalized : "-";
}

export function severityColor(value?: string | null) {
  const normalized = normalizeSeverityLabel(value);
  return normalized ? severityColors[normalized] ?? "default" : "default";
}

export function isHighOrCriticalSeverity(value?: string | null) {
  return ["critical", "high"].includes(normalizeSeverityLabel(value) ?? "");
}

function compactKey(value: string) {
  return value.trim().toLowerCase().replace(/[\s/_\-.]+/g, "");
}
